### ISSUE TITLE
[ML] Turn around the Runner's class hierarchy

### ISSUE THREAD

**Issue by valeriy42** [elastic/ml-cpp#735]
At the moment the class hierarchy for `CDataFrameAnalysisRunner` classes looks as following
```
CDataFrameAnalysisRunner <|-- CDataFrameOutliersRunner
CDataFrameAnalysisRunner <|-- CDataFrameBoostedTreeRunner
CDataFrameBoostedTreeRunner <|-- CDataFrameClassificationRunner
CDataFrameBoostedTreeRunner <|-- CDataFrameRegressionRunner
```
It seems to me that since `CDataFrameClassificationRunner` and `CDataFrameRegressionRunner` signal the problem we are solving (same as `CDataFrameOutliersRunner`), while `CDataFrameBoostedTreeRunner` denote a particular ML algorithm used for solving these problems, we will run into a problem wehn we will implement another ML algorithm (e.g. neural networks).

Hence, I suggest to turn the class hierarchy around and let `CDataFrameBoostedTreeRunner` depend on `CDataFrameClassificationRunner` and `CDataFrameRegressionRunner`. 

@tveasey WDYT?

**Comment by tveasey** [comment]
I have decided on reflection I'm not completely happy with the naming either. I had two thoughts, which will depend on how we intend to incorporate a new model type:
1. `CDataFrameBoostedTreeRunner` -> `CDataFrameTrainPredictiveModelRunner`, `CDataFrameClassificationRunner` -> `CDataFrameTrainClassifierRunner` and `CDataFrameRegressionRunner` -> `CDataFrameTrainRegressionRunner`
2. `CDataFrameBoostedTreeRunner` ->`CDataFrameTrainBoostedTreeRunner`, `CDataFrameClassificationRunner` -> `CDataFrameTrainBoostedTreeClassifierRunner` and
`CDataFrameRegressionRunner` -> `CDataFrameTrainBoostedTreeRegressionRunner`.

This depends on whether we're planning on using polymorphism in the `CDataFrameAnalysisRunner` hierarchy or `CDataFrameRegressionModel`. I'm coming round to the view point that we should go with the former, because, at the least, all the overridable parameters will be learner specific. (In which case, we should probably drop `CDataFrameRegressionModel`.)

We could definitely consider turning things around too. The downside I see is that there is then not a route to easily share the common code for the same algorithm through inheritance, but there are probably alternative ways of achieving this.

**Comment by valeriy42** [comment]
Will it work if we make `CBoostedTreeRunner` to a `pimpl` for `CDataFrameRegressionRunner` and `CDataFrameClassificationRunner`? IMO, It would make sense using aggregation in this case, since boosted tree **is** an implementation of the regression or classification abstractions.

**Comment by tveasey** [comment]
Yes it does feel like composition could also work for this case. Also, I'm in agreement that we need to do something with names. I want to see how we can handle say the parameter reading for this approach. My inclination would be to try it out and see what it looks like.
